### PR TITLE
Add clang style check as part of the sanity check

### PR DESCRIPTION
**Pull request by yongtang**
This fix is an effort to add clang style check as part of the sanity check.

**NOTE: Not sure if this PR is relevant so feel free to close it if it doesn't make sense**

In `CONTRIBUTING.md` it has been advised to run `clang-format --style=google file.cc` so that Google coding style is conformed. However, there is no sanity check in the current Jenkins build so current .cc and .h files in the repo are not really conforming to the coding style.

This actually causes issues. In case a PR is submitted with `clang-format --style=google file.cc`, the reviewer may see additional unrelated changes which might be a distraction. The developer may also spent additional time to manually check for any discrepancies manually with additional unrelated style changes.

This fix adds the clang-format check to the ci build so that when `ci_sanity.sh` is running, it will use clang-format to make sure the code is conforming to the coding style as specified in `CONTRIBUTING.md`.

One thing that might need to take notice is the header order of the Eigen library. See
https://github.com/tensorflow/tensorflow/pull/13907#issuecomment-338718110 for further details.

Basically, if Eigen headers could be placed in any order, then no additional steps are needed. Otherwise, it is always possible to place the Eigen headers at the top, then leave one empty line like:
```cpp
#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"

#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/op_kernel.h"
...
```

In this way, even a run of `clang-format -i --style=google file.cc` will still respect the order and leave Eigen header at the top.

This PR is experimeal so it only checks `tensorflow/core/ops` directory. Other files could be added if this PR is OK.

This PR also sanitizes all files in `tensorflow/core/ops` directory so that it conforms to coding style requirement.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by yifeif** [comment]
Thanks for the PR @yongtang, this will be very helpful! cc: @gunan

**Comment by gunan** [comment]
Jenkins, test this please.

File: tensorflow/tools/ci_build/clang_format_check.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# limitations under the License.
+# ==============================================================================
+
+# please run this at root directory of tensorflow
+success=1
+
+CLANG_FORMAT=${CLANG_FORMAT:-clang-format}
+
+# only tensorflow/core/ops is checked at the moment for experimental purpose
+for filename in $(find tensorflow/core/ops -name *.h -o -name *.cc); do
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow/tools/ci_build/clang_format_check.sh]: Thanks. Done.

**Comment by yongtang** [comment]
Thanks @yifeif @gunan, the PR has been updated with incremental options added. Please take a look and let me know if there are any issues.

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by yongtang** [comment]
@yifeif The previous Jenkins error was caused by the fact that full check (instead of `--incremental`) was done on Jenkins' `Sanity Checks`.

I have updated the PR with all `.cc` and `.h` files sanitized with `clang-format -i --style=Google` though a script I wrote.

I also checked the header and make sure all `third_party/eigen3` headers remains in place (without recording). That could be verified through (no output is expected if all order remains):
```sh
$ git diff HEAD~XXX | grep -E '#include "third_party/eigen3' | grep -E '^[-+]'
```

I think the PR will help the community to spend less time on style check and more time on implantation.

The PR was split into multiple commits for easy review. It could be squashed later.

Please take a look and let me know if there are any issues.

**Comment by benoitsteiner** [comment]
Can we add the sanity checks to ensure new PRs are correctly formatted, but grandfather the existing style violations ? That will simplify the merging of this PR into our internal repository by reducing the number of merge conflicts to deal with. If needs be, we can always gradually fix the formatting in several smaller PRs spread over time.

**Comment by yongtang** [comment]
Thanks @benoitsteiner for the suggestion. The gradual strategy makes sense and it should help avoiding the merge conflict.

I have stripped out the commits with style format changes, and enabled `--incremental` by default. The `--incremental` could be removed after all files are touched over time.

Please take a look and let me know if there are any issues.

**Review by yifeif — APPROVED** [review]

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by benoitsteiner** [comment]
Jenkins, test this please.

File: tensorflow/tools/ci_build/ci_sanity.sh [hunk 2]
```diff
@@ -439,7 +505,7 @@ do_check_load_py_test() {
 }
 
 # Supply all sanity step commands and descriptions
-SANITY_STEPS=("do_pylint PYTHON2" "do_pylint PYTHON3" "do_buildifier" "do_bazel_nobuild" "do_pip_package_licenses_check" "do_lib_package_licenses_check" "do_java_package_licenses_check" "do_pip_smoke_test" "do_check_load_py_test" "do_code_link_check")
+SANITY_STEPS=("do_pylint PYTHON2" "do_pylint PYTHON3" "do_buildifier" "do_bazel_nobuild" "do_pip_package_licenses_check" "do_lib_package_licenses_check" "do_java_package_licenses_check" "do_pip_smoke_test" "do_check_load_py_test" "do_code_link_check" "do_clang_format_check")
 SANITY_STEPS_DESC=("Python 2 pylint" "Python 3 pylint" "buildifier check" "bazel nobuild" "pip: license check for external dependencies" "C library: license check for external dependencies" "Java Native Library: license check for external dependencies" "Pip Smoke Test: Checking py_test dependencies exist in pip package" "Check load py_test: Check that BUILD files with py_test target properly load py_test" "Code Link Check: Check there are no broken links")
```
Comment (caisq) [inline tensorflow/tools/ci_build/ci_sanity.sh]: Please add a corresponding description string for `do_clang_format_check` to `SANITY_STEPS_DESC`

Comment (yongtang) [reply tensorflow/tools/ci_build/ci_sanity.sh]: Thanks @caisq. A follow up PR #14030 has been created with `SANITY_STEPS_DESC` added. Please take a look.
